sys/usb/usbus/usbus.c

```diff
@@ -0,0 +1,667 @@
+    usbdev_t *dev = usbus->dev;
+    usbus->pid = sched_active_pid;
+    usbus->addr = 0;
+    usbus->strings = NULL;
+    usbus->iface = NULL;
+    usbus->str_idx = 1;
+    usbus->setup_state = USBUS_SETUPRQ_READY;
+    mutex_init(&usbus->lock);
+    msg_t msg, msg_queue[_USBUS_MSG_QUEUE_SIZE];
+    DEBUG("usbus: starting thread %i\n", sched_active_pid);
+    /* setup the link-layer's message queue */
+    msg_init_queue(msg_queue, _USBUS_MSG_QUEUE_SIZE);
+    /* register the event callback with the device driver */
+    dev->cb = _event_cb;
+    /* initialize low-level driver */
+    dev->driver->init(dev);
+    dev->context = usbus;
+
+    usbus->in = usbus->dev->driver->new_ep(usbus->dev, USB_EP_TYPE_CONTROL, USB_EP_DIR_IN, USBUS_MAX_SIZE);
+    usbus->out = usbus->dev->driver->new_ep(usbus->dev, USB_EP_TYPE_CONTROL, USB_EP_DIR_OUT, USBUS_MAX_SIZE);
+    usbus->in->cb = _event_ep0_cb;
+    usbus->out->cb = _event_ep0_cb;
+    usbus->in->context = usbus;
+    usbus->out->context = usbus;
+
+    usbus->in->driver->init(usbus->in);
+    usbus->out->driver->init(usbus->out);
+    _usbus_config_ep0(usbus);
+    usbus_add_string_descriptor(usbus, &usbus->config, USB_CONFIG_CONFIGURATION_STR);
+    usbus_add_string_descriptor(usbus, &usbus->product, USB_CONFIG_PRODUCT_STR);
+    usbus_add_string_descriptor(usbus, &usbus->manuf, USB_CONFIG_MANUF_STR);
+
+    usbus->state = USBUS_STATE_DISCONNECT;
+    mutex_unlock(&usbus->lock);
+
+    /* TODO: signal ready from application */
+    xtimer_sleep(1);
+    usbopt_enable_t enable = USBOPT_ENABLE;
+    dev->driver->set(dev, USBOPT_ATTACH, &enable, sizeof(usbopt_enable_t));
+
+    while (1) {
+        msg_receive(&msg);
+        /* dispatch netdev, MAC and gnrc_netapi messages */
+        switch (msg.type) {
+            case USBUS_MSG_TYPE_EVENT:
+                dev->driver->esr(dev);
+                break;
+            case USBUS_MSG_TYPE_EP_EVENT:
+                {
+                    usbdev_ep_t *ep = (usbdev_ep_t*)msg.content.ptr;
+                    ep->driver->esr(ep);
+                }
+              break;
+            default:
+                DEBUG("usbus: unhandled event\n");
+                break;
+        }
+    }
+    return NULL;
+}
+
+/* USB event callback */
+void _event_cb(usbdev_t *usbdev, usbdev_event_t event)
+{
+    usbus_t *usbus = (usbus_t *)usbdev->context;
+    if (event == USBDEV_EVENT_ESR) {
+        msg_t msg = { .type = USBUS_MSG_TYPE_EVENT,
+                      .content = { .ptr = usbdev } };
+
+        if (msg_send(&msg, usbus->pid) <= 0) {
+            puts("usbus: possibly lost interrupt.");
+        }
+    }
+    else {
+        switch (event) {
+            case USBDEV_EVENT_RESET:
+                {
+                usbus->state = USBUS_STATE_RESET;
+                usbus->addr = 0;
+                usbus->setup_state = USBUS_SETUPRQ_READY;
+                usbus->dev->driver->set(usbus->dev, USBOPT_ADDRESS, &usbus->addr, sizeof(uint8_t));
```

Comment:
Is the `sizeof` correct? I'm asking because `usbus->addr` is a `uint16_t`.
And if your answer is no, then I would strongly suggest to change it to `sizeof(usbus->addr)`
And if your answer is yes, you should realize you have an endianess problem.

Reply:
Not that it is breaking anything at the moment, but `usbus->addr` should be `uint8_t` :confused:

Reply:
Even if the `addr` should be `uint8_t`, it is safer to program this as `sizeof(usbus->addr)`.

Reply:
Agreed, fixed